src/SonarQube.Client/Services/SonarQubeClient.cs

```diff
@@ -0,0 +1,324 @@
+﻿/*
+ * SonarQube Client
+ * Copyright (C) 2016-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web;
+using Google.Protobuf;
+using Newtonsoft.Json.Linq;
+using SonarQube.Client.Helpers;
+using SonarQube.Client.Messages;
+
+namespace SonarQube.Client.Services
+{
+    public sealed class SonarQubeClient : ISonarQubeClient, IDisposable
+    {
+        private readonly HttpClient httpClient;
+        private bool isDisposed;
+
+        public SonarQubeClient(ConnectionRequest connection, HttpMessageHandler messageHandler, TimeSpan requestTimeout)
+        {
+            if (connection == null)
+            {
+                throw new ArgumentNullException(nameof(connection));
+            }
+            if (messageHandler == null)
+            {
+                throw new ArgumentNullException(nameof(messageHandler));
+            }
+            if (requestTimeout <= TimeSpan.Zero)
+            {
+                throw new ArgumentException("Doesn't expect a zero or negative timeout.", nameof(requestTimeout));
+            }
+
+            this.httpClient = new HttpClient(messageHandler)
+            {
+                BaseAddress = connection.ServerUri,
+                Timeout = requestTimeout
+            };
+            this.httpClient.DefaultRequestHeaders.Authorization = AuthenticationHeaderFactory.Create(connection);
+        }
+
+        public void Dispose()
+        {
+            if (isDisposed)
+            {
+                return;
+            }
+
+            this.httpClient.Dispose();
+            this.isDisposed = true;
+        }
+
+        public Task<Result<ComponentResponse[]>> GetComponentsSearchProjectsAsync(ComponentRequest request,
+            CancellationToken token)
+        {
+            // Since 6.2; internal
+            var apiPath = BuildRelativeUrl("api/components/search_projects",
+                new Dictionary<string, string>
+                {
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this 'TODO' comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)